### ISSUE TITLE
Duality using a different SynchronizationContext inside the editor

### ISSUE THREAD

**Issue by Barsonax** [AdamsLair/duality#592]
### Summary

Currently inside the editor duality uses the [WindowsFormsSynchronizationContext](https://msdn.microsoft.com/en-us/library/system.windows.forms.windowsformssynchronizationcontext(v=vs.110).aspx). However outside the editor it uses no SynchronizationContext (which means this context is either null or equal to [SynchronizationContext](https://msdn.microsoft.com/en-us/library/system.threading.synchronizationcontext(v=vs.110).aspx). This can cause code to behave in a different way when running outside the editor and most likely it will break if this is the case.

To fix this inconsistency the following solutions could be applied:
1. Make sure that duality uses no SynchronizationContext inside the editor. However since winforms depends on this this will probably mean you need to run winforms and duality in separate threads. I suspect many things will break if this is done.
2. Make a custom DualitySynchronizationContext that will invoke the continuations in the OnUpdate loop. When using await this will make Tasks have the same behavior as my [PathRequest](https://github.com/Barsonax/Pathfindax/blob/develop/Source/Code/Pathfindax/PathfindEngine/PathRequest.cs) class in Pathfindax meaning continuations will execute on the mainthread. Make sure this context is only applied outside the editor as it will probably break winforms.
3. Document this is a difference between running inside the editor and outside.

I suspect 1 would be the most complex since it has huge impacts all over the editor followed by 2 and 3.

### How to reproduce

- Create a new component and implement ICmpInitializable
- Put the following code in the OnInit method (the following code could also be done with await syntax):
```csharp
				Log.Game.Write($"Init thread: {Thread.CurrentThread.ManagedThreadId} Context: {SynchronizationContext.Current}");
				Task.Run(() =>
				{
					Log.Game.Write($"Task run Thread: {Thread.CurrentThread.ManagedThreadId}");
				}).ContinueWith(x =>
				{
					Log.Game.Write($"Task run continuation Thread: {Thread.CurrentThread.ManagedThreadId}");
				}, TaskScheduler.FromCurrentSynchronizationContext());
```
- Run this code from inside the editor and from visual studio. Note that inside the editor the continuation is on the mainthread but in visual studio its on a different thread (a threadpool thread to be precise).

### Workaround
- While possible it can be quite alot of work to workaround this. In pathfindax I made a custom awaitable with its own custom logic so that the continuation is synced to the mainthread thus avoiding this whole issue.

### Analysis
If going for 1:
- Duality wil run isolated from any UI no matter if its inside an editor or not.
- Likely very complex to implement requiring many changes.

If going for 2:
- Proper support for Tasks as the continuation will execute on the mainthread. Basically the same behavior when running a winforms or WPF application.

If going for 3:
- Basically saying duality does not properly support Tasks as it can lead to inconsistent behavior.
- Easy to implement.

**Comment by ilexp** [comment]
**Note:** Some potentially related research regarding synchronization contexts can also be found in issue #295. Might be worth looking at as well, with regard to avoiding deadlock potential in manual synchronization adjustments.

**Comment by Barsonax** [comment]
In pathfindax iam passing the synchronization context as a constructor parameter [here](https://github.com/Barsonax/Pathfindax/blob/master/Source/Code/Pathfindax/PathfindEngine/PathfindaxManager.cs) instead of it being static (note that I also added interfaces). I do think such a approach would require a custom awaitable though. On the other hand it removes the evil static state.
